README.md
# OpenAI Code Generator GitHub Action 🚀

## Overview 📖
This GitHub Action allows you to generate code files using the OpenAI API. It's perfect for automating the process of file generation and directly integrating these files into your codebase. After generation, it commits the files and optionally opens a pull request.

## Features 🌟
 - Code Generation: Generates files based on given descriptions.
 - Commit & Push: Commits the generated files and pushes to a new or existing branch.
 - Pull Requests: Optionally creates a pull request against a base branch.
 - Dry Run: Test the workflow without making any changes to the repo.

## Usage 🛠️
Include the action in your GitHub Actions workflow YAML:

```yaml
- name: OpenAI Code Generator
  uses: https://github.com/JuliusHaring/llm-code-gen@main
  with:
    descriptions: 'python linting with flake8 and github actions'
    openai_token: ${{ secrets.OPENAI_TOKEN }}
```


## Contributions 🤝
Feel free to open an issue or submit a pull request if you want to contribute.

## License 📝
This project is licensed under the MIT License. See the LICENSE file for details.

Enjoy using the OpenAI Code Generator GitHub Action! 🎉

CHANGELOG.md
# CHANGELOG



## v1.0.0 (2023-09-11)

### Chore

* chore: rename action ([`275c958`](https://github.com/JuliusHaring/llm-code-gen/commit/275c95874978fe589b17ec3e433b53e022fa9ff0))

* chore: rename job step ([`a2201ae`](https://github.com/JuliusHaring/llm-code-gen/commit/a2201ae3fbf3b7c364407f0a3155c4fa615f8e66))

### Documentation

* docs: add license and readme ([`5688c53`](https://github.com/JuliusHaring/llm-code-gen/commit/5688c53f57df804325ad5a44fd4adfb05418369b))

### Refactor

* refactor: add inputs to run action dispatch ([`b790575`](https://github.com/JuliusHaring/llm-code-gen/commit/b7905752cbe9aca74c5e99ceb14d49bf628e7a10))


## v0.4.0 (2023-09-11)

### Feature

* feat: add pr configuration ([`8018784`](https://github.com/JuliusHaring/llm-code-gen/commit/80187841cb3302ac18d3058a00eb75442f90c253))


## v0.3.0 (2023-09-11)

### Feature

* feat: add ability to create PR ([`82601b1`](https://github.com/JuliusHaring/llm-code-gen/commit/82601b1e5b4f0e484fe5b0baa8dea1d3f7d20697))

### Unknown

* Merge pull request #4 from JuliusHaring/create-pr

feat: add ability to create PR ([`1a22bb6`](https://github.com/JuliusHaring/llm-code-gen/commit/1a22bb616c9eda59272af861e3aaabc7e03b523d))


## v0.2.0 (2023-09-11)

### Chore

* chore: adjust debugging ([`d43ffe4`](https://github.com/JuliusHaring/llm-code-gen/commit/d43ffe439c83e7e35ae3a58ebad4cc7d533f33b1))

* chore: adjust prompt ([`efa359b`](https://github.com/JuliusHaring/llm-code-gen/commit/efa359b6cddc670550fb3f33e3d0aafb29936d96))

* chore: adjust prompt ([`8f697e7`](https://github.com/JuliusHaring/llm-code-gen/commit/8f697e7cca0a2662a4896c64967aa27a0e68d3f1))

### Ci

* ci: add test to ci script ([`69181a1`](https://github.com/JuliusHaring/llm-code-gen/commit/69181a110f1c2f61feba26d1e61ec73b6d8739bc))

* ci: add test to cd ([`8af0262`](https://github.com/JuliusHaring/llm-code-gen/commit/8af0262d2ecc96286f66b8e95aa468b38270600a))

### Feature

* feat: add custom commit msg ([`43cbce9`](https://github.com/JuliusHaring/llm-code-gen/commit/43cbce9e76b6558024d65db236dd90abdc8810ee))

* feat: introduce retries and dict validation ([`ab3982b`](https://github.com/JuliusHaring/llm-code-gen/commit/ab3982b7a44247920add6ae402cd92fb31dd5bce))

* feat: add force push option ([`f8a34ab`](https://github.com/JuliusHaring/llm-code-gen/commit/f8a34abefafae0ff462b2c6ad7f5a7674c4167b9))

### Fix

* fix: update gh token env variable ([`5d0e966`](https://github.com/JuliusHaring/llm-code-gen/commit/5d0e966d37094d161d9f6acd058ed737345be115))

### Refactor

* refactor: outsource git logic to action ([`7eeae97`](https://github.com/JuliusHaring/llm-code-gen/commit/7eeae97d80cf872f18b150047728143fcf3da101))

* refactor: outsource codes to functions ([`1210aef`](https://github.com/JuliusHaring/llm-code-gen/commit/1210aef250b34ab7279f3796b5c289ee75b6c751))

### Unknown

* Merge pull request #2 from JuliusHaring/py-enhancements

Py enhancements ([`a5183d9`](https://github.com/JuliusHaring/llm-code-gen/commit/a5183d9d0878600b82409ecabd00eb3688361a70))


## v0.1.0 (2023-09-10)

### Ci

* ci: add cd script ([`fe560e6`](https://github.com/JuliusHaring/llm-code-gen/commit/fe560e6eecb33aeb6d0f49febc5427e57614b927))

* ci: add ci script ([`54de13c`](https://github.com/JuliusHaring/llm-code-gen/commit/54de13c8142bc181481f7acdf8e6164a0c6e2735))

### Feature

* feat: make file generation aware of all files ([`bd6f222`](https://github.com/JuliusHaring/llm-code-gen/commit/bd6f22211918799523b9f079330c2a24a02ccd7d))

* feat: add workflow_dispatch to test ([`5b070df`](https://github.com/JuliusHaring/llm-code-gen/commit/5b070df45f2e413538b52841d61926c04c0cb1f1))

* feat: add log_level as arg ([`a4f53d7`](https://github.com/JuliusHaring/llm-code-gen/commit/a4f53d76951c468f8bc95c9117eadb6e9e365ec7))

* feat: add max_tokens as input ([`14dcd51`](https://github.com/JuliusHaring/llm-code-gen/commit/14dcd51c4fe5debb4092b8d7f8fd16cd3447fb2f))

### Fix

* fix: fix cd dependencies ([`a81e1b3`](https://github.com/JuliusHaring/llm-code-gen/commit/a81e1b3de617a4cb9c00e994783fac70f76135ed))

### Refactor

* refactor: rename code to suite usage ([`816c2ce`](https://github.com/JuliusHaring/llm-code-gen/commit/816c2ceb357eb26c27d6ad11ba18905db41d9a9b))

* refactor: rename files to suite usage ([`120e150`](https://github.com/JuliusHaring/llm-code-gen/commit/120e150899a805f39a852edaacd386a5676d3cbb))

### Unknown

* Merge pull request #1 from JuliusHaring/ci

ci: add ci script ([`aa20f32`](https://github.com/JuliusHaring/llm-code-gen/commit/aa20f3237cd8b983647bbf0f1836094794f7ee40))

* run linting / formatting ([`5bef621`](https://github.com/JuliusHaring/llm-code-gen/commit/5bef6219b88375f8aef0f233cf5fa462d664b1ae))

* fix initial bugs ([`6b3e70e`](https://github.com/JuliusHaring/llm-code-gen/commit/6b3e70e66cec55f4f3b281949a0efbcab1ec113f))

* add test ([`18c5839`](https://github.com/JuliusHaring/llm-code-gen/commit/18c583958bf3eeb1184d97d327603ac7cb02951a))

* initial commit ([`adc4094`](https://github.com/JuliusHaring/llm-code-gen/commit/adc409439df0fb3d2bd9c3a75e1944c439383bb9))


generate_ci_cd.py
import os
import openai
import json
import logging
from git import Repo
import argparse

logging.basicConfig(level=logging.INFO)

def ensure_directory(file_path):
    directory = os.path.dirname(file_path)
    if directory and not os.path.exists(directory):
        os.makedirs(directory)
    logging.info(f"Ensured directory for {file_path}.")

def openai_call(prompt, model_name, max_tokens=300, role="user"):
    messages = [
        {"role": "system", "content": "You are a script that helps set up a Repository with initial files."},
        {"role": role, "content": prompt}
    ]
    response = openai.ChatCompletion.create(model=model_name, messages=messages, max_tokens=max_tokens)
    logging.info("Retrieved result:\n%s", response)
    message_content = response.choices[0].message['content'].strip()
    logging.info("Returning message:\n%s", message_content)
    return message_content

def generate_ci_cd(branch_name, keywords, openai_token, model_name):
    openai.api_key = openai_token

    prompt_for_files = f"List the files and their specific directories and requirements for a CI/CD setup with the following keywords: {keywords}. Provide the information as a compact, JSON-formatted dictionary where the keys are file paths (not directories, just paths!) and the values are lists of keywords. Make sure the output is parseable by Python's json.loads."
    files_and_requirements = json.loads(openai_call(prompt_for_files, model_name))
    logging.info(f"Generated file and directory requirements: {files_and_requirements}")

    repo = Repo('.')

    remote_url = repo.remotes.origin.url
    new_url = remote_url.replace("https://", f"https://{os.getenv('GITHUB_TOKEN')}:x-oauth-basic@")
    repo.remotes.origin.set_url(new_url)

    logging.info("Remote URL set to: %s", repo.remotes.origin.url)


    new_branch = repo.create_head(branch_name)
    new_branch.checkout()
    logging.info(f"Checked out new branch: {branch_name}")

    for file_path, file_keywords in files_and_requirements.items():
        ensure_directory(file_path)

        prompt_for_content = f"Generate content for {file_path} with the following keywords: {file_keywords}."
        file_content = openai_call(prompt_for_content, model_name)
        logging.info(f"Generated content for {file_path}.")

        with open(file_path, "w") as f:
            f.write(file_content)
    
    repo.git.add(all=True)
    repo.index.commit("Add generated CI/CD setup")
    repo.git.push("origin", branch_name)
    logging.info(f"Committed and pushed changes to {branch_name}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Generate CI/CD setup.")
    parser.add_argument('--branch_name', default="ci-cd-setup", help="Name of the new branch.")
    parser.add_argument('--keywords', required=True, help="Keywords to inform the setup.")
    parser.add_argument('--openai_token', required=True, help="OpenAI API Token.")
    parser.add_argument('--model_name', default="gpt-3.5-turbo-16k-0613", help="OpenAI model name.")
    
    args = parser.parse_args()
    generate_ci_cd(args.branch_name, args.keywords, args.openai_token, args.model_name)


generate_files.py
import argparse
import json
import logging
import os

import openai


def ensure_directory(file_path):
    directory = os.path.dirname(file_path)
    if directory and not os.path.exists(directory):
        os.makedirs(directory)
    logging.info(f"Ensured directory for {file_path}.")


def validate_dict_output(raw_output):
    logging.info("Validating dict:\n%s", str(raw_output))
    try:
        parsed_output = json.loads(raw_output)
        if not isinstance(parsed_output, dict):
            logging.error("Output is not a dictionary.")
            return False, None
        for key, value in parsed_output.items():
            if not isinstance(key, str) or not isinstance(value, str):
                logging.error(
                    "Dictionary keys and/or values are not strings: %s, %s.",
                    str(key),
                    str(value),
                )
                return False, None
        logging.info("Validation passed.")
        return True, parsed_output
    except json.JSONDecodeError:
        logging.error("Failed to decode JSON.")
        return False, None


def openai_call(prompt, model_name, max_tokens=300):
    logging.info("Running prompt with model: %s", model_name)
    logging.debug("Using prompt:\n%s", prompt)
    messages = [
        {
            "role": "system",
            "content": "You are a script that helps set up a Repository with initial files.",
        },
        {"role": "user", "content": prompt},
    ]
    logging.debug("Using messages:\n%s", messages)
    response = openai.ChatCompletion.create(
        model=model_name, messages=messages, max_tokens=max_tokens
    )
    logging.debug(f"Retrieved result:\n{response}")
    return response.choices[0].message["content"].strip()


def get_file_requirements_dict(descriptions, model_name, max_tokens):
    prompt_for_files = (
        "List the files and their specific directories and requirements for a"
        f"production ready codebase using the following descriptions: '{descriptions}'."
        "Stick to that description, don't add anything else. Use the minimum amount of required files."
        "Make the minimum amount of assumptions about the rest of the repo."
        "Provide the information as a compact, JSON-formatted dictionary where the keys are file paths"
        "(not directories, just files!) and the values are the descriptions strings."
    )
    raw_output = openai_call(prompt_for_files, model_name, max_tokens)
    is_valid, files_and_requirements = validate_dict_output(raw_output)
    if is_valid:
        logging.debug(
            f"Generated file and directory requirements: {files_and_requirements}"
        )
    return is_valid, files_and_requirements


def get_file_content(
    file_path, file_descriptions, files_and_requirements, model_name, max_tokens
):
    prompt_for_content = (
        "Here is a list of all files to be generated and their requirements:"
        "\n-----------\n"
        f" {json.dumps(files_and_requirements)}"
        "\n-----------\n"
        f"Create the file content of '{file_path}' from the following description: {file_descriptions}. "
        "Only output the content for this specific file."
        "Your output will be directly copied to the file, so you will output proper content,"
        "no comments about it."
        f"Content of the file '{file_path}':"
    )
    file_content = openai_call(prompt_for_content, model_name, max_tokens)
    logging.info(f"Generated content for {file_path}.")
    return file_content


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Generate code using OpenAI API.")
    parser.add_argument("--descriptions", required=True)
    parser.add_argument("--openai_token", required=True)
    parser.add_argument("--model_name", default="gpt-3.5-turbo-16k-0613")
    parser.add_argument("--max_tokens", default=300, type=int)
    parser.add_argument("--log_level", default="INFO", type=str)
    parser.add_argument("--force_push", action="store_true", default=True)

    args = parser.parse_args()
    logging.basicConfig(level=args.log_level)
    openai.api_key = args.openai_token

    is_valid, files_and_requirements = get_file_requirements_dict(
        args.descriptions, args.model_name, args.max_tokens
    )

    if not is_valid:
        logging.error("Invalid file requirements dictionary. Exiting.")
        exit(1)

    for file_path, file_descriptions in files_and_requirements.items():
        if ".github" in file_path:
            logging.info("Replacing .github in %s", file_path)
            file_path = file_path.replace(".github", "github")

        if os.path.exists(file_path):
            logging.warning(f"File {file_path} already exists. Skipping.")
            continue

        ensure_directory(file_path)
        file_content = get_file_content(
            file_path,
            file_descriptions,
            files_and_requirements,
            args.model_name,
            args.max_tokens,
        )

        with open(file_path, "w") as f:
            logging.info("Writing to: %s", file_path)
            f.write(file_content)
